src/getPublicKeys.c

```diff
@@ -0,0 +1,284 @@
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(HANDLE_INIT_UI_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		ctx->stage = GET_KEYS_STAGE_GET_KEYS;
+	}
+	UI_STEP_END(HANDLE_INIT_UI_STEP_INVALID);
+}
+
+static void getPublicKeys_handleInitAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		CHECK_STAGE(GET_KEYS_STAGE_INIT);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		// initialization
+		ctx->remainingKeys = 0;
+	}
+	{
+		// parse data
+
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		struct {
+			uint8_t numKeys[4];
+		}* wireHeader = (void*) wireDataBuffer;
+
+		VALIDATE(sizeof(*wireHeader) == wireDataSize, ERR_INVALID_DATA);
+
+		uint32_t remainingKeys = u4be_read(wireHeader->numKeys);
+		TRACE("num keys: %d", remainingKeys);
+		VALIDATE(remainingKeys <= MAX_PUBLIC_KEYS, ERR_INVALID_DATA);
+		VALIDATE(remainingKeys > 0, ERR_INVALID_DATA);
+
+		ASSERT_TYPE(ctx->remainingKeys, uint16_t);
+		ASSERT(ctx->remainingKeys <= UINT16_MAX);
+		ctx->remainingKeys = (uint16_t) remainingKeys;
+	}
+
+	security_policy_t policy = policyForGetPublicKeysInit();
+	ENSURE_NOT_DENIED(policy);
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {ctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_PROMPT_BEFORE_RESPONSE, HANDLE_INIT_UI_STEP_CONFIRM);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT,   HANDLE_INIT_UI_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+
+	getPublicKeys_handleInit_ui_runStep();
+}
+
+
+// ============================== GET KEY HANDLER ==============================
+
+
+void getPublicKeys_handleGetKeyAPDU(
+        uint8_t *wireDataBuffer,
+        size_t wireDataSize
+)
+{
+    CHECK_STAGE(GET_KEYS_STAGE_GET_KEYS);
+
+    ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+
+    VALIDATE(ctx->remainingKeys >= 1, ERR_INVALID_STATE);
+    ctx->remainingKeys--;
+
+	returnOneKey(wireDataBuffer, wireDataSize);
+}
+
+// ============================== GET SINGLE KEY HANDLER ==============================
+
+void getPublicKeys_handleGetSingleKeyAPDU(
+        uint8_t *wireDataBuffer,
+        size_t wireDataSize
+)
+{
+	// only allow this as a new call, not in the middle of a bulk key export
+    CHECK_STAGE(GET_KEYS_STAGE_INIT);
+
+    ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+
+	returnOneKey(wireDataBuffer, wireDataSize);
+}
+
+// ============================== MAIN HANDLER ==============================
+
+typedef void subhandler_fn_t(uint8_t* dataBuffer, size_t dataSize);
+
+static subhandler_fn_t* lookup_subhandler(uint8_t p1)
+{
+	switch(p1) {
+#	define  CASE(P1, HANDLER) case P1: return HANDLER;
+#	define  DEFAULT(HANDLER)  default: return HANDLER;
+		// legacy mechanism for getting a single key
+		CASE(P1_UNUSED, getPublicKeys_handleGetSingleKeyAPDU);
+
+		CASE(0x01, getPublicKeys_handleInitAPDU);
+		CASE(0x02, getPublicKeys_handleGetKeyAPDU);
+		DEFAULT(NULL)
+#	undef   CASE
+#	undef   DEFAULT
+	}
+}
+
+void getPublicKeys_handleAPDU(
```

Comment:
is this function used anywhere?

Reply:
Yes, it's the main entry point used in `handlers.c`.